invalid memory address or nil pointer dereference

**What did you do?**

I ran the latest build from master, and it paniced.

**What did you expect to see?**

Happy Prometheus.

**What did you see instead? Under which circumstances?**

```
level=info ts=2018-01-04T12:34:30.797481431Z caller=main.go:221 msg="Starting Prometheus" version="(version=2.0.0, branch=master, revision=ec94df49d45baa8e680d81e732294eb04fff00de)"
level=info ts=2018-01-04T12:34:30.797587605Z caller=main.go:222 build_context="(go=go1.9.2, user=root@1885792ecf46, date=20171229-20:03:34)"
level=info ts=2018-01-04T12:34:30.797618333Z caller=main.go:223 host_details="(Linux 4.4.64+ #1 SMP Wed Aug 30 20:27:36 PDT 2017 x86_64 weave-cortex-agent-1352299515-q9b40 (none))"
level=info ts=2018-01-04T12:34:30.797643323Z caller=main.go:224 fd_limits="(soft=1048576, hard=1048576)"
level=info ts=2018-01-04T12:34:30.802483848Z caller=manager.go:58 component="scrape manager" msg="Starting scrape manager..."
level=info ts=2018-01-04T12:34:30.802561868Z caller=main.go:434 msg="Starting TSDB ..."
level=info ts=2018-01-04T12:34:30.806440473Z caller=web.go:383 component=web msg="Start listening for connections" address=:8080
level=info ts=2018-01-04T12:34:30.812878794Z caller=main.go:444 msg="TSDB started"
level=info ts=2018-01-04T12:34:30.812978586Z caller=main.go:511 msg="Loading configuration file" filename=/etc/prometheus/prometheus.yml
level=info ts=2018-01-04T12:34:30.814840891Z caller=kubernetes.go:191 component="discovery manager" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2018-01-04T12:34:30.816859216Z caller=kubernetes.go:191 component="discovery manager" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2018-01-04T12:34:30.817553956Z caller=kubernetes.go:191 component="discovery manager" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2018-01-04T12:34:30.818090353Z caller=kubernetes.go:191 component="discovery manager" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2018-01-04T12:34:30.81978125Z caller=kubernetes.go:191 component="discovery manager" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2018-01-04T12:34:30.821172572Z caller=main.go:421 msg="Server is ready to receive requests."
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x155d32a]

goroutine 123 [running]:
github.com/prometheus/prometheus/discovery.(*Manager).allGroups.func1(0x29062c0, 0xc420402c00)
	/go/src/github.com/prometheus/prometheus/discovery/manager.go:188 +0x28a
github.com/prometheus/prometheus/discovery.(*Manager).Run(0xc4203a9700, 0x29062c0, 0xc420402c00, 0x0, 0x0)
	/go/src/github.com/prometheus/prometheus/discovery/manager.go:96 +0x54
main.main.func5(0x0, 0x0)
	/go/src/github.com/prometheus/prometheus/cmd/prometheus/main.go:338 +0x52
github.com/prometheus/prometheus/vendor/github.com/oklog/oklog/pkg/group.(*Group).Run.func1(0xc420052240, 0xc42047fb60, 0xc4204424c0)
	/go/src/github.com/prometheus/prometheus/vendor/github.com/oklog/oklog/pkg/group/group.go:38 +0x27
created by github.com/prometheus/prometheus/vendor/github.com/oklog/oklog/pkg/group.(*Group).Run
	/go/src/github.com/prometheus/prometheus/vendor/github.com/oklog/oklog/pkg/group/group.go:37 +0xa8
```

**Environment**

Kubernetes 1.7 on GKE.

* Prometheus version:

```
> docker run prom/prometheus@sha256:6050e90c156d4e6be31a23e87b53673b2a2d2ab5b71677225555675b43a1b445 --version                 
prometheus, version 2.0.0 (branch: master, revision: ec94df49d45baa8e680d81e732294eb04fff00de)
  build user:       root@1885792ecf46
  build date:       20171229-20:03:34
  go version:       go1.9.2
```

* Prometheus configuration file:
```
        global:
          scrape_interval: 5s
        remote_write:
          - url: 'https://cloud.weave.works./api/prom/push'
            basic_auth:
              password: <snip>
        remote_read:
          - url: 'https://cloud.weave.works./api/prom/read'
            basic_auth:
              password: <snip>
        scrape_configs:
          - job_name: kubernetes-apiservers
            kubernetes_sd_configs:
              - role: endpoints
            scheme: https
            tls_config:
              ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
            bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
            relabel_configs:
              - source_labels:
                  - __meta_kubernetes_namespace
                  - __meta_kubernetes_service_name
                  - __meta_kubernetes_endpoint_port_name
                action: keep
                regex: default;kubernetes;https
              - source_labels:
                  - __meta_kubernetes_namespace
                target_label: _weave_namespace
              - source_labels:
                  - __meta_kubernetes_endpoints_name
                target_label: _weave_service
          - job_name: kubernetes-pods
            kubernetes_sd_configs:
              - role: pod
            relabel_configs:
              - source_labels:
                  - __meta_kubernetes_pod_annotation_prometheus_io_scrape
                action: drop
                regex: 'false'
              - source_labels:
                  - __meta_kubernetes_pod_annotation_prometheus_io_scheme
                action: replace
                target_label: __scheme__
                regex: ^(https?)$
                replacement: $1
              - source_labels:
                  - __meta_kubernetes_pod_annotation_prometheus_io_path
                action: replace
                target_label: __metrics_path__
                regex: ^(.+)$
                replacement: $1
              - source_labels:
                  - __address__
                  - __meta_kubernetes_pod_annotation_prometheus_io_port
                action: replace
                target_label: __address__
                regex: '([^:]+)(?::\d+)?;(\d+)'
                replacement: '$1:$2'
              - source_labels:
                  - __meta_kubernetes_namespace
                target_label: _weave_namespace
              - source_labels:
                  - __meta_kubernetes_pod_name
                  - __meta_kubernetes_pod_node_name
                target_label: node
                regex: ^weave-cortex-node-exporter-.+;(.+)$
                replacement: $1
              - source_labels:
                  - _weave_service
                  - __meta_kubernetes_pod_name
                target_label: _weave_service
                regex: '^;(kube-.*)-(?:ip|gke)-.*$'
                replacement: $1
              - source_labels:
                  - _weave_service
                  - __meta_kubernetes_pod_name
                target_label: _weave_service
                regex: '^;(.*?)(?:(?:-[0-9bcdf]+)?-[0-9a-z]{5})?$'
                replacement: $1
              - source_labels:
                  - _weave_service
                  - __meta_kubernetes_pod_name
                regex: ^;(.+)$
                target_label: _weave_service
                replacement: $1
          - job_name: kubernetes-nodes
            kubernetes_sd_configs:
              - role: node
            tls_config:
              insecure_skip_verify: true
            bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
            relabel_configs:
              - target_label: __scheme__
                replacement: https
              - target_label: __address__
                replacement: 'kubernetes.default.svc:443'
              - source_labels:
                  - __meta_kubernetes_node_name
                regex: (.+)
                target_label: __metrics_path__
                replacement: '/api/v1/nodes/${1}/proxy/metrics'
              - target_label: _weave_namespace
                replacement: default
              - target_label: _weave_service
                replacement: kubelet
          - job_name: weave-net
            kubernetes_sd_configs:
              - role: pod
            relabel_configs:
              - source_labels:
                  - __meta_kubernetes_namespace
                  - __meta_kubernetes_pod_label_name
                action: keep
                regex: ^kube-system;weave-net$
              - source_labels:
                  - __meta_kubernetes_pod_container_name
                  - __address__
                action: replace
                target_label: __address__
                regex: '^weave;(.+?)(?::\d+)?$'
                replacement: '$1:6782'
              - source_labels:
                  - __meta_kubernetes_pod_container_name
                  - __address__
                action: replace
                target_label: __address__
                regex: '^weave-npc;(.+?)(?::\d+)?$'
                replacement: '$1:6781'
              - source_labels:
                  - __meta_kubernetes_namespace
                target_label: _weave_namespace
              - target_label: _weave_service
                replacement: weave-net
          - job_name: cadvisor
            kubernetes_sd_configs:
              - role: node
            tls_config:
              insecure_skip_verify: true
            bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
            scheme: https
            relabel_configs:
              - target_label: __address__
                replacement: 'kubernetes.default.svc:443'
              - source_labels:
                  - __meta_kubernetes_node_name
                regex: (.+)
                target_label: __metrics_path__
                replacement: '/api/v1/nodes/${1}/proxy/metrics/cadvisor'
              - target_label: _weave_namespace
                replacement: default
              - target_label: _weave_service
                replacement: cadvisor
            metric_relabel_configs:
              - source_labels:
                  - _weave_pod_name
                  - pod_name
                target_label: _weave_pod_name
                regex: '^;(kube-.*)-(?:ip|gke)-.*$'
                replacement: $1
              - source_labels:
                  - _weave_pod_name
                  - pod_name
                target_label: _weave_pod_name
                regex: '^;(.*?)(?:(?:-[0-9bcdf]+)?-[0-9a-z]{5})?$'
                replacement: $1
              - source_labels:
                  - _weave_pod_name
                  - pod_name
                regex: ^;(.+)$
                target_label: _weave_pod_name
                replacement: $1
```

@brian-brazil I see you have referenced an updated release 2.1.0 in the comments above.

We have encountered the same panic (below) from the latest master release which we are currently troubleshooting:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x155d32a]
```

We are using the :master release tagged quay.io/prometheus/prometheus SHA256: 0b745814b3c5

Do you have an approximate ETA for when the 2.10 release may be available that may resolve this issue?

--- a/discovery/manager.go
+++ b/discovery/manager.go
@@ -69,8 +69,8 @@ func NewManager(logger log.Logger) *Manager {
 	}
 }
 
-// Manager maintains a set of discovery providers and sends each update to a channel used by other packages.
-// Targets sent to the channel are grouped by the target set name.
+// Manager maintains a set of discovery providers and sends each update to a map channel.
+// Targets are grouped by the target set name.
 type Manager struct {
 	logger         log.Logger
 	actionCh       chan func(context.Context)
@@ -182,7 +182,7 @@ func (m *Manager) allGroups() map[string][]*targetgroup.Group {
 		for pkey, tsets := range m.targets {
 			for _, tg := range tsets {
 				// Even if the target group 'tg' is empty we still need to send it to the 'Scrape manager'
-				// to singal that is needs to stop all scrape loops for this target set.
+				// to signal that it needs to stop all scrape loops for this target set.
 				tSetsAll[pkey.setName] = append(tSetsAll[pkey.setName], tg)
 			}
 		}

--- a/discovery/manager.go
+++ b/discovery/manager.go
@@ -155,11 +155,6 @@ func (m *Manager) updateGroup(poolKey poolKey, tgs []*targetgroup.Group) {
 	done := make(chan struct{})
 
 	m.actionCh <- func(ctx context.Context) {
-		if tgs == nil {
-			close(done)
-			return
-		}
-
 		for _, tg := range tgs {
 			if tg != nil { // Some Discoverers send nil target group so need to check for it to avoid panics.
 				if _, ok := m.targets[poolKey]; !ok {

--- a/discovery/manager.go
+++ b/discovery/manager.go
@@ -73,7 +73,7 @@ func NewManager(logger log.Logger) *Manager {
 		logger:         logger,
 		actionCh:       make(chan func(context.Context)),
 		syncCh:         make(chan map[string][]*targetgroup.Group),
-		targets:        make(map[poolKey][]*targetgroup.Group),
+		targets:        make(map[poolKey]map[string]*targetgroup.Group),
 		discoverCancel: []context.CancelFunc{},
 	}
 }
@@ -83,7 +83,8 @@ type Manager struct {
 	logger         log.Logger
 	actionCh       chan func(context.Context)
 	discoverCancel []context.CancelFunc
-	targets        map[poolKey][]*targetgroup.Group
+	// We use map[string]*targetgroup.Group to handle Discoverers that send only updates instead of all targets on every update.
+	targets map[poolKey]map[string]*targetgroup.Group
 	// The sync channels sends the updates in map[targetSetName] where targetSetName is the job value from the scrape config.
 	syncCh chan map[string][]*targetgroup.Group
 }
@@ -143,8 +144,8 @@ func (m *Manager) runProvider(ctx context.Context, poolKey poolKey, updates chan
 			if !ok {
 				return
 			}
-			m.addGroup(poolKey, tgs)
-			m.syncCh <- m.allGroups(poolKey)
+			m.updateGroup(poolKey, tgs)
+			m.syncCh <- m.allGroups()
 		}
 	}
 }
@@ -153,16 +154,21 @@ func (m *Manager) cancelDiscoverers() {
 	for _, c := range m.discoverCancel {
 		c()
 	}
-	m.targets = make(map[poolKey][]*targetgroup.Group)
+	m.targets = make(map[poolKey]map[string]*targetgroup.Group)
 	m.discoverCancel = nil
 }
 
-func (m *Manager) addGroup(poolKey poolKey, tg []*targetgroup.Group) {
+func (m *Manager) updateGroup(poolKey poolKey, tg []*targetgroup.Group) {
 	done := make(chan struct{})
 
 	m.actionCh <- func(ctx context.Context) {
 		if tg != nil {
-			m.targets[poolKey] = tg
+			for _, t := range tg {
+				if _, ok := m.targets[poolKey]; !ok {
+					m.targets[poolKey] = make(map[string]*targetgroup.Group)
+				}
+				m.targets[poolKey][t.Source] = t
+			}
 		}
 		close(done)
 
@@ -170,7 +176,7 @@ func (m *Manager) addGroup(poolKey poolKey, tg []*targetgroup.Group) {
 	<-done
 }
 
-func (m *Manager) allGroups(pk poolKey) map[string][]*targetgroup.Group {
+func (m *Manager) allGroups() map[string][]*targetgroup.Group {
 	tSets := make(chan map[string][]*targetgroup.Group)
 
 	m.actionCh <- func(ctx context.Context) {
@@ -185,7 +191,9 @@ func (m *Manager) allGroups(pk poolKey) map[string][]*targetgroup.Group {
 		tSetsAll := map[string][]*targetgroup.Group{}
 		for _, pk := range pKeys {
 			for _, tg := range m.targets[pk] {
-				if tg.Source != "" { // Don't add empty targets.
+				// Don't add empty targets.
+				// Some Discoverers(eg. k8s) send only the updates so removed targets will be updated with an empty Source value.
+				if tg.Source != "" {
 					tSetsAll[pk.setName] = append(tSetsAll[pk.setName], tg)
 				}
 			}

--- a/discovery/manager.go
+++ b/discovery/manager.go
@@ -70,6 +70,8 @@ func NewManager(logger log.Logger) *Manager {
 }
 
 // Manager maintains a set of discovery providers and sends each update to a channel used by other packages.
+// Targets are grouped by target set names.
+// When a given target set doesn't include any targets the manager doesn't send any updates for this target set.
 type Manager struct {
 	logger         log.Logger
 	actionCh       chan func(context.Context)
@@ -160,7 +162,7 @@ func (m *Manager) updateGroup(poolKey poolKey, tgs []*targetgroup.Group) {
 		}
 
 		for _, tg := range tgs {
-			if tg != nil { // Some Discoverers send nil targetgroup so need to check for it to avoid panics.
+			if tg != nil { // Some Discoverers send nil target group so need to check for it to avoid panics.
 				if _, ok := m.targets[poolKey]; !ok {
 					m.targets[poolKey] = make(map[string]*targetgroup.Group)
 				}
@@ -181,6 +183,8 @@ func (m *Manager) allGroups() map[string][]*targetgroup.Group {
 		for pkey, tsets := range m.targets {
 			del := true
 			for _, tg := range tsets {
+				// Don't add a target set if the target group has no targets.
+				// This happens when a Discoverer sends an empty targets array for a group to indicate that these targets have been removed.
 				if len(tg.Targets) != 0 {
 					tSetsAll[pkey.setName] = append(tSetsAll[pkey.setName], tg)
 					del = false

--- a/discovery/manager.go
+++ b/discovery/manager.go
@@ -16,7 +16,6 @@ package discovery
 import (
 	"context"
 	"fmt"
-	"sort"
 
 	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
@@ -58,14 +57,6 @@ type poolKey struct {
 	setName  string
 	provider string
 }
-
-// byProvider implements sort.Interface for []poolKey based on the provider field.
-// Sorting is needed so that we can have predictable tests.
-type byProvider []poolKey
-
-func (a byProvider) Len() int           { return len(a) }
-func (a byProvider) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
-func (a byProvider) Less(i, j int) bool { return a[i].provider < a[j].provider }
 
 // NewManager is the Discovery Manager constructor
 func NewManager(logger log.Logger) *Manager {
@@ -182,18 +173,10 @@ func (m *Manager) allGroups() map[string][]*targetgroup.Group {
 	tSets := make(chan map[string][]*targetgroup.Group)
 
 	m.actionCh <- func(ctx context.Context) {
-
-		// Sorting by the poolKey is needed so that we can have predictable tests.
-		var pKeys []poolKey
-		for pk := range m.targets {
-			pKeys = append(pKeys, pk)
-		}
-		sort.Sort(byProvider(pKeys))
-
 		tSetsAll := map[string][]*targetgroup.Group{}
-		for _, pk := range pKeys {
-			for _, tg := range m.targets[pk] {
-				tSetsAll[pk.setName] = append(tSetsAll[pk.setName], tg)
+		for pkey, tsets := range m.targets {
+			for _, tg := range tsets {
+				tSetsAll[pkey.setName] = append(tSetsAll[pkey.setName], tg)
 			}
 		}
 		tSets <- tSetsAll

--- a/discovery/manager.go
+++ b/discovery/manager.go
@@ -75,7 +75,7 @@ type Manager struct {
 	actionCh       chan func(context.Context)
 	discoverCancel []context.CancelFunc
 	// Some Discoverers(eg. k8s) send only the updates for a given target group
-	// so we use map[tg.Source]*targetgroup.Group to  know which group to update.
+	// so we use map[tg.Source]*targetgroup.Group to know which group to update.
 	targets map[poolKey]map[string]*targetgroup.Group
 	// The sync channels sends the updates in map[targetSetName] where targetSetName is the job value from the scrape config.
 	syncCh chan map[string][]*targetgroup.Group

--- a/discovery/manager.go
+++ b/discovery/manager.go
@@ -154,14 +154,17 @@ func (m *Manager) updateGroup(poolKey poolKey, tgs []*targetgroup.Group) {
 	done := make(chan struct{})
 
 	m.actionCh <- func(ctx context.Context) {
-		if tgs != nil {
-			for _, tg := range tgs {
-				if tg != nil { // Some Discoverers send nil targetgroup so need to check for it to avoid panics.
-					if _, ok := m.targets[poolKey]; !ok {
-						m.targets[poolKey] = make(map[string]*targetgroup.Group)
-					}
-					m.targets[poolKey][tg.Source] = tg
+		if tgs == nil {
+			close(done)
+			return
+		}
+
+		for _, tg := range tgs {
+			if tg != nil { // Some Discoverers send nil targetgroup so need to check for it to avoid panics.
+				if _, ok := m.targets[poolKey]; !ok {
+					m.targets[poolKey] = make(map[string]*targetgroup.Group)
 				}
+				m.targets[poolKey][tg.Source] = tg
 			}
 		}
 		close(done)

--- a/discovery/manager.go
+++ b/discovery/manager.go
@@ -179,8 +179,16 @@ func (m *Manager) allGroups() map[string][]*targetgroup.Group {
 	m.actionCh <- func(ctx context.Context) {
 		tSetsAll := map[string][]*targetgroup.Group{}
 		for pkey, tsets := range m.targets {
+			del := true
 			for _, tg := range tsets {
-				tSetsAll[pkey.setName] = append(tSetsAll[pkey.setName], tg)
+				if len(tg.Targets) != 0 {
+					tSetsAll[pkey.setName] = append(tSetsAll[pkey.setName], tg)
+					del = false
+				}
+			}
+			// Delete the empty map for this target set to avoid memory leaks.
+			if del {
+				delete(m.targets, pkey)
 			}
 		}
 		tSets <- tSetsAll

--- a/discovery/manager.go
+++ b/discovery/manager.go
@@ -74,7 +74,8 @@ type Manager struct {
 	logger         log.Logger
 	actionCh       chan func(context.Context)
 	discoverCancel []context.CancelFunc
-	// We use map[string]*targetgroup.Group to handle Discoverers that send only updates instead of all targets on every update.
+	// Some Discoverers(eg. k8s) send only the updates for a given target group
+	// so we use map[tg.Source]*targetgroup.Group to  know which group to update.
 	targets map[poolKey]map[string]*targetgroup.Group
 	// The sync channels sends the updates in map[targetSetName] where targetSetName is the job value from the scrape config.
 	syncCh chan map[string][]*targetgroup.Group

--- a/discovery/manager.go
+++ b/discovery/manager.go
@@ -70,8 +70,7 @@ func NewManager(logger log.Logger) *Manager {
 }
 
 // Manager maintains a set of discovery providers and sends each update to a channel used by other packages.
-// Targets are grouped by target set names.
-// When a given target set doesn't include any targets the manager doesn't send any updates for this target set.
+// Targets sent to the channel are grouped by the target set name.
 type Manager struct {
 	logger         log.Logger
 	actionCh       chan func(context.Context)
@@ -181,18 +180,10 @@ func (m *Manager) allGroups() map[string][]*targetgroup.Group {
 	m.actionCh <- func(ctx context.Context) {
 		tSetsAll := map[string][]*targetgroup.Group{}
 		for pkey, tsets := range m.targets {
-			del := true
 			for _, tg := range tsets {
-				// Don't add a target set if the target group has no targets.
-				// This happens when a Discoverer sends an empty targets array for a group to indicate that these targets have been removed.
-				if len(tg.Targets) != 0 {
-					tSetsAll[pkey.setName] = append(tSetsAll[pkey.setName], tg)
-					del = false
-				}
-			}
-			// Delete the empty map for this target set to avoid memory leaks.
-			if del {
-				delete(m.targets, pkey)
+				// Even if the target group 'tg' is empty we still need to send it to the 'Scrape manager'
+				// to singal that is needs to stop all scrape loops for this target set.
+				tSetsAll[pkey.setName] = append(tSetsAll[pkey.setName], tg)
 			}
 		}
 		tSets <- tSetsAll

--- a/discovery/manager.go
+++ b/discovery/manager.go
@@ -158,16 +158,18 @@ func (m *Manager) cancelDiscoverers() {
 	m.discoverCancel = nil
 }
 
-func (m *Manager) updateGroup(poolKey poolKey, tg []*targetgroup.Group) {
+func (m *Manager) updateGroup(poolKey poolKey, tgs []*targetgroup.Group) {
 	done := make(chan struct{})
 
 	m.actionCh <- func(ctx context.Context) {
-		if tg != nil {
-			for _, t := range tg {
-				if _, ok := m.targets[poolKey]; !ok {
-					m.targets[poolKey] = make(map[string]*targetgroup.Group)
+		if tgs != nil {
+			for _, tg := range tgs {
+				if tg != nil { // Some Discoverers send nil targetgroup so need to check for it to avoid panics.
+					if _, ok := m.targets[poolKey]; !ok {
+						m.targets[poolKey] = make(map[string]*targetgroup.Group)
+					}
+					m.targets[poolKey][tg.Source] = tg
 				}
-				m.targets[poolKey][t.Source] = t
 			}
 		}
 		close(done)
@@ -191,11 +193,7 @@ func (m *Manager) allGroups() map[string][]*targetgroup.Group {
 		tSetsAll := map[string][]*targetgroup.Group{}
 		for _, pk := range pKeys {
 			for _, tg := range m.targets[pk] {
-				// Don't add empty targets.
-				// Some Discoverers(eg. k8s) send only the updates so removed targets will be updated with an empty Source value.
-				if tg.Source != "" {
-					tSetsAll[pk.setName] = append(tSetsAll[pk.setName], tg)
-				}
+				tSetsAll[pk.setName] = append(tSetsAll[pk.setName], tg)
 			}
 		}
 		tSets <- tSetsAll

--- a/discovery/manager_test.go
+++ b/discovery/manager_test.go
@@ -17,6 +17,7 @@ import (
 	"context"
 	"fmt"
 	"reflect"
+	"sort"
 	"strconv"
 	"testing"
 	"time"
@@ -103,11 +104,11 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "initial1",
+								Source:  "tp1_group1",
 								Targets: []model.LabelSet{{"__instance__": "1"}},
 							},
 							{
-								Source:  "initial2",
+								Source:  "tp1_group2",
 								Targets: []model.LabelSet{{"__instance__": "2"}},
 							}},
 					},
@@ -116,11 +117,11 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 			expectedTargets: [][]*targetgroup.Group{
 				{
 					{
-						Source:  "initial1",
+						Source:  "tp1_group1",
 						Targets: []model.LabelSet{{"__instance__": "1"}},
 					},
 					{
-						Source:  "initial2",
+						Source:  "tp1_group2",
 						Targets: []model.LabelSet{{"__instance__": "2"}},
 					},
 				},
@@ -133,11 +134,11 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "tp1-initial1",
+								Source:  "tp1_group1",
 								Targets: []model.LabelSet{{"__instance__": "1"}},
 							},
 							{
-								Source:  "tp1-initial2",
+								Source:  "tp1_group2",
 								Targets: []model.LabelSet{{"__instance__": "2"}},
 							},
 						},
@@ -147,7 +148,7 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "tp2-initial1",
+								Source:  "tp2_group1",
 								Targets: []model.LabelSet{{"__instance__": "3"}},
 							},
 						},
@@ -158,24 +159,24 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 			expectedTargets: [][]*targetgroup.Group{
 				{
 					{
-						Source:  "tp1-initial1",
+						Source:  "tp1_group1",
 						Targets: []model.LabelSet{{"__instance__": "1"}},
 					},
 					{
-						Source:  "tp1-initial2",
+						Source:  "tp1_group2",
 						Targets: []model.LabelSet{{"__instance__": "2"}},
 					},
 				}, {
 					{
-						Source:  "tp1-initial1",
+						Source:  "tp1_group1",
 						Targets: []model.LabelSet{{"__instance__": "1"}},
 					},
 					{
-						Source:  "tp1-initial2",
+						Source:  "tp1_group2",
 						Targets: []model.LabelSet{{"__instance__": "2"}},
 					},
 					{
-						Source:  "tp2-initial1",
+						Source:  "tp2_group1",
 						Targets: []model.LabelSet{{"__instance__": "3"}},
 					},
 				},
@@ -188,48 +189,11 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "initial1",
+								Source:  "tp1_group1",
 								Targets: []model.LabelSet{{"__instance__": "1"}},
 							},
 							{
-								Source:  "initial2",
-								Targets: []model.LabelSet{{"__instance__": "2"}},
-							},
-						},
-						interval: 0,
-					},
-					{
-						targetGroups: []targetgroup.Group{},
-						interval:     10,
-					},
-				},
-			},
-			expectedTargets: [][]*targetgroup.Group{
-				{
-					{
-						Source:  "initial1",
-						Targets: []model.LabelSet{{"__instance__": "1"}},
-					},
-					{
-						Source:  "initial2",
-						Targets: []model.LabelSet{{"__instance__": "2"}},
-					},
-				},
-				{},
-			},
-		},
-		{
-			title: "Single TP initials and new groups",
-			updates: map[string][]update{
-				"tp1": {
-					{
-						targetGroups: []targetgroup.Group{
-							{
-								Source:  "initial1",
-								Targets: []model.LabelSet{{"__instance__": "1"}},
-							},
-							{
-								Source:  "initial2",
+								Source:  "tp1_group2",
 								Targets: []model.LabelSet{{"__instance__": "2"}},
 							},
 						},
@@ -238,12 +202,12 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "update1",
-								Targets: []model.LabelSet{{"__instance__": "3"}},
+								Source:  "tp1_group1",
+								Targets: []model.LabelSet{},
 							},
 							{
-								Source:  "update2",
-								Targets: []model.LabelSet{{"__instance__": "4"}},
+								Source:  "tp1_group2",
+								Targets: []model.LabelSet{},
 							},
 						},
 						interval: 10,
@@ -253,22 +217,85 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 			expectedTargets: [][]*targetgroup.Group{
 				{
 					{
-						Source:  "initial1",
+						Source:  "tp1_group1",
 						Targets: []model.LabelSet{{"__instance__": "1"}},
 					},
 					{
-						Source:  "initial2",
+						Source:  "tp1_group2",
 						Targets: []model.LabelSet{{"__instance__": "2"}},
 					},
 				},
 				{
 					{
-						Source:  "update1",
+						Source:  "tp1_group1",
+						Targets: []model.LabelSet{},
+					},
+					{
+						Source:  "tp1_group2",
+						Targets: []model.LabelSet{},
+					},
+				},
+			},
+		},
+		{
+			title: "Single TP initials and new groups",
+			updates: map[string][]update{
+				"tp1": {
+					{
+						targetGroups: []targetgroup.Group{
+							{
+								Source:  "tp1_group1",
+								Targets: []model.LabelSet{{"__instance__": "1"}},
+							},
+							{
+								Source:  "tp1_group2",
+								Targets: []model.LabelSet{{"__instance__": "2"}},
+							},
+						},
+						interval: 0,
+					},
+					{
+						targetGroups: []targetgroup.Group{
+							{
+								Source:  "tp1_group1",
+								Targets: []model.LabelSet{{"__instance__": "3"}},
+							},
+							{
+								Source:  "tp1_group2",
+								Targets: []model.LabelSet{{"__instance__": "4"}},
+							},
+							{
+								Source:  "tp1_group3",
+								Targets: []model.LabelSet{{"__instance__": "1"}},
+							},
+						},
+						interval: 10,
+					},
+				},
+			},
+			expectedTargets: [][]*targetgroup.Group{
+				{
+					{
+						Source:  "tp1_group1",
+						Targets: []model.LabelSet{{"__instance__": "1"}},
+					},
+					{
+						Source:  "tp1_group2",
+						Targets: []model.LabelSet{{"__instance__": "2"}},
+					},
+				},
+				{
+					{
+						Source:  "tp1_group1",
 						Targets: []model.LabelSet{{"__instance__": "3"}},
 					},
 					{
-						Source:  "update2",
+						Source:  "tp1_group2",
 						Targets: []model.LabelSet{{"__instance__": "4"}},
+					},
+					{
+						Source:  "tp1_group3",
+						Targets: []model.LabelSet{{"__instance__": "1"}},
 					},
 				},
 			},
@@ -280,11 +307,11 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "tp1-initial1",
+								Source:  "tp1_group1",
 								Targets: []model.LabelSet{{"__instance__": "1"}},
 							},
 							{
-								Source:  "tp1-initial2",
+								Source:  "tp1_group2",
 								Targets: []model.LabelSet{{"__instance__": "2"}},
 							},
 						},
@@ -293,11 +320,11 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "tp1-update1",
+								Source:  "tp1_group3",
 								Targets: []model.LabelSet{{"__instance__": "3"}},
 							},
 							{
-								Source:  "tp1-update2",
+								Source:  "tp1_group4",
 								Targets: []model.LabelSet{{"__instance__": "4"}},
 							},
 						},
@@ -308,11 +335,11 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "tp2-initial1",
+								Source:  "tp2_group1",
 								Targets: []model.LabelSet{{"__instance__": "5"}},
 							},
 							{
-								Source:  "tp2-initial2",
+								Source:  "tp2_group2",
 								Targets: []model.LabelSet{{"__instance__": "6"}},
 							},
 						},
@@ -321,11 +348,11 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "tp2-update1",
+								Source:  "tp2_group3",
 								Targets: []model.LabelSet{{"__instance__": "7"}},
 							},
 							{
-								Source:  "tp2-update2",
+								Source:  "tp2_group4",
 								Targets: []model.LabelSet{{"__instance__": "8"}},
 							},
 						},
@@ -336,82 +363,106 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 			expectedTargets: [][]*targetgroup.Group{
 				{
 					{
-						Source:  "tp1-initial1",
+						Source:  "tp1_group1",
 						Targets: []model.LabelSet{{"__instance__": "1"}},
 					},
 					{
-						Source:  "tp1-initial2",
+						Source:  "tp1_group2",
 						Targets: []model.LabelSet{{"__instance__": "2"}},
 					},
 				},
 				{
 					{
-						Source:  "tp1-initial1",
+						Source:  "tp1_group1",
 						Targets: []model.LabelSet{{"__instance__": "1"}},
 					},
 					{
-						Source:  "tp1-initial2",
+						Source:  "tp1_group2",
 						Targets: []model.LabelSet{{"__instance__": "2"}},
 					},
 					{
-						Source:  "tp2-initial1",
+						Source:  "tp2_group1",
 						Targets: []model.LabelSet{{"__instance__": "5"}},
 					},
 					{
-						Source:  "tp2-initial2",
+						Source:  "tp2_group2",
 						Targets: []model.LabelSet{{"__instance__": "6"}},
 					},
 				},
 				{
 					{
-						Source:  "tp1-initial1",
+						Source:  "tp1_group1",
 						Targets: []model.LabelSet{{"__instance__": "1"}},
 					},
 					{
-						Source:  "tp1-initial2",
+						Source:  "tp1_group2",
 						Targets: []model.LabelSet{{"__instance__": "2"}},
 					},
 					{
-						Source:  "tp2-update1",
+						Source:  "tp2_group1",
+						Targets: []model.LabelSet{{"__instance__": "5"}},
+					},
+					{
+						Source:  "tp2_group2",
+						Targets: []model.LabelSet{{"__instance__": "6"}},
+					},
+					{
+						Source:  "tp2_group3",
 						Targets: []model.LabelSet{{"__instance__": "7"}},
 					},
 					{
-						Source:  "tp2-update2",
+						Source:  "tp2_group4",
 						Targets: []model.LabelSet{{"__instance__": "8"}},
 					},
 				},
 				{
 					{
-						Source:  "tp1-update1",
+						Source:  "tp1_group1",
+						Targets: []model.LabelSet{{"__instance__": "1"}},
+					},
+					{
+						Source:  "tp1_group2",
+						Targets: []model.LabelSet{{"__instance__": "2"}},
+					},
+					{
+						Source:  "tp1_group3",
 						Targets: []model.LabelSet{{"__instance__": "3"}},
 					},
 					{
-						Source:  "tp1-update2",
+						Source:  "tp1_group4",
 						Targets: []model.LabelSet{{"__instance__": "4"}},
 					},
 					{
-						Source:  "tp2-update1",
+						Source:  "tp2_group1",
+						Targets: []model.LabelSet{{"__instance__": "5"}},
+					},
+					{
+						Source:  "tp2_group2",
+						Targets: []model.LabelSet{{"__instance__": "6"}},
+					},
+					{
+						Source:  "tp2_group3",
 						Targets: []model.LabelSet{{"__instance__": "7"}},
 					},
 					{
-						Source:  "tp2-update2",
+						Source:  "tp2_group4",
 						Targets: []model.LabelSet{{"__instance__": "8"}},
 					},
 				},
 			},
 		},
 		{
-			title: "One tp initials arrive after other tp updates.",
+			title: "One TP initials arrive after other TP updates.",
 			updates: map[string][]update{
 				"tp1": {
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "tp1-initial1",
+								Source:  "tp1_group1",
 								Targets: []model.LabelSet{{"__instance__": "1"}},
 							},
 							{
-								Source:  "tp1-initial2",
+								Source:  "tp1_group2",
 								Targets: []model.LabelSet{{"__instance__": "2"}},
 							},
 						},
@@ -420,11 +471,11 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "tp1-update1",
+								Source:  "tp1_group1",
 								Targets: []model.LabelSet{{"__instance__": "3"}},
 							},
 							{
-								Source:  "tp1-update2",
+								Source:  "tp1_group2",
 								Targets: []model.LabelSet{{"__instance__": "4"}},
 							},
 						},
@@ -435,11 +486,11 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "tp2-initial1",
+								Source:  "tp2_group1",
 								Targets: []model.LabelSet{{"__instance__": "5"}},
 							},
 							{
-								Source:  "tp2-initial2",
+								Source:  "tp2_group2",
 								Targets: []model.LabelSet{{"__instance__": "6"}},
 							},
 						},
@@ -448,11 +499,11 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "tp2-update1",
+								Source:  "tp2_group1",
 								Targets: []model.LabelSet{{"__instance__": "7"}},
 							},
 							{
-								Source:  "tp2-update2",
+								Source:  "tp2_group2",
 								Targets: []model.LabelSet{{"__instance__": "8"}},
 							},
 						},
@@ -463,57 +514,57 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 			expectedTargets: [][]*targetgroup.Group{
 				{
 					{
-						Source:  "tp1-initial1",
+						Source:  "tp1_group1",
 						Targets: []model.LabelSet{{"__instance__": "1"}},
 					},
 					{
-						Source:  "tp1-initial2",
+						Source:  "tp1_group2",
 						Targets: []model.LabelSet{{"__instance__": "2"}},
 					},
 				},
 				{
 					{
-						Source:  "tp1-update1",
+						Source:  "tp1_group1",
 						Targets: []model.LabelSet{{"__instance__": "3"}},
 					},
 					{
-						Source:  "tp1-update2",
+						Source:  "tp1_group2",
 						Targets: []model.LabelSet{{"__instance__": "4"}},
 					},
 				},
 				{
 					{
-						Source:  "tp1-update1",
+						Source:  "tp1_group1",
 						Targets: []model.LabelSet{{"__instance__": "3"}},
 					},
 					{
-						Source:  "tp1-update2",
+						Source:  "tp1_group2",
 						Targets: []model.LabelSet{{"__instance__": "4"}},
 					},
 					{
-						Source:  "tp2-initial1",
+						Source:  "tp2_group1",
 						Targets: []model.LabelSet{{"__instance__": "5"}},
 					},
 					{
-						Source:  "tp2-initial2",
+						Source:  "tp2_group2",
 						Targets: []model.LabelSet{{"__instance__": "6"}},
 					},
 				},
 				{
 					{
-						Source:  "tp1-update1",
+						Source:  "tp1_group1",
 						Targets: []model.LabelSet{{"__instance__": "3"}},
 					},
 					{
-						Source:  "tp1-update2",
+						Source:  "tp1_group2",
 						Targets: []model.LabelSet{{"__instance__": "4"}},
 					},
 					{
-						Source:  "tp2-update1",
+						Source:  "tp2_group1",
 						Targets: []model.LabelSet{{"__instance__": "7"}},
 					},
 					{
-						Source:  "tp2-update2",
+						Source:  "tp2_group2",
 						Targets: []model.LabelSet{{"__instance__": "8"}},
 					},
 				},
@@ -521,34 +572,43 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 		},
 
 		{
-			title: "Single TP Single provider empty update in between",
+			title: "Single TP empty update in between",
 			updates: map[string][]update{
 				"tp1": {
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "initial1",
+								Source:  "tp1_group1",
 								Targets: []model.LabelSet{{"__instance__": "1"}},
 							},
 							{
-								Source:  "initial2",
+								Source:  "tp1_group2",
 								Targets: []model.LabelSet{{"__instance__": "2"}},
 							},
 						},
 						interval: 30,
 					},
 					{
-						targetGroups: []targetgroup.Group{},
-						interval:     10,
+						targetGroups: []targetgroup.Group{
+							{
+								Source:  "tp1_group1",
+								Targets: []model.LabelSet{},
+							},
+							{
+								Source:  "tp1_group2",
+								Targets: []model.LabelSet{},
+							},
+						},
+						interval: 10,
 					},
 					{
 						targetGroups: []targetgroup.Group{
 							{
-								Source:  "update1",
+								Source:  "tp1_group1",
 								Targets: []model.LabelSet{{"__instance__": "3"}},
 							},
 							{
-								Source:  "update2",
+								Source:  "tp1_group2",
 								Targets: []model.LabelSet{{"__instance__": "4"}},
 							},
 						},
@@ -559,22 +619,31 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 			expectedTargets: [][]*targetgroup.Group{
 				{
 					{
-						Source:  "initial1",
+						Source:  "tp1_group1",
 						Targets: []model.LabelSet{{"__instance__": "1"}},
 					},
 					{
-						Source:  "initial2",
+						Source:  "tp1_group2",
 						Targets: []model.LabelSet{{"__instance__": "2"}},
 					},
 				},
-				{},
 				{
 					{
-						Source:  "update1",
+						Source:  "tp1_group1",
+						Targets: []model.LabelSet{},
+					},
+					{
+						Source:  "tp1_group2",
+						Targets: []model.LabelSet{},
+					},
+				},
+				{
+					{
+						Source:  "tp1_group1",
 						Targets: []model.LabelSet{{"__instance__": "3"}},
 					},
 					{
-						Source:  "update2",
+						Source:  "tp1_group2",
 						Targets: []model.LabelSet{{"__instance__": "4"}},
 					},
 				},
@@ -606,6 +675,8 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 				break Loop
 			case tsetMap := <-discoveryManager.SyncCh():
 				for _, received := range tsetMap {
+					// Need to sort by the Groups source as the Discovery manager doesn't guarantee the order.
+					sort.Sort(byGroupSource(received))
 					if !reflect.DeepEqual(received, testCase.expectedTargets[x]) {
 						var receivedFormated string
 						for _, receivedTargets := range received {
@@ -628,7 +699,7 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 }
 
 func TestTargetSetRecreatesTargetGroupsEveryRun(t *testing.T) {
-	verifyPresence := func(tSets map[poolKey][]*targetgroup.Group, poolKey poolKey, label string, present bool) {
+	verifyPresence := func(tSets map[poolKey]map[string]*targetgroup.Group, poolKey poolKey, label string, present bool) {
 		if _, ok := tSets[poolKey]; !ok {
 			t.Fatalf("'%s' should be present in Pool keys: %v", poolKey, tSets)
 			return
@@ -729,3 +800,10 @@ func (tp mockdiscoveryProvider) sendUpdates() {
 		tp.up <- tgs
 	}
 }
+
+// byGroupSource implements sort.Interface so we can sort by the Source field.
+type byGroupSource []*targetgroup.Group
+
+func (a byGroupSource) Len() int           { return len(a) }
+func (a byGroupSource) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byGroupSource) Less(i, j int) bool { return a[i].Source < a[j].Source }

--- a/discovery/manager_test.go
+++ b/discovery/manager_test.go
@@ -649,57 +649,6 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 				},
 			},
 		},
-		{
-			title: "Single TP update with an empty group to check for memory leaks",
-			updates: map[string][]update{
-				"tp1": {
-					{
-						targetGroups: []targetgroup.Group{
-							{
-								Source:  "tp1_group1",
-								Targets: []model.LabelSet{{"__instance__": "1"}},
-							},
-							{
-								Source:  "tp1_group2",
-								Targets: []model.LabelSet{{"__instance__": "2"}},
-							},
-						},
-						interval: 30,
-					},
-					{
-						targetGroups: []targetgroup.Group{
-							{
-								Source:  "tp1_group1",
-								Targets: []model.LabelSet{{"__instance__": "3"}},
-							},
-							{
-								Source:  "tp1_group2",
-								Targets: nil,
-							},
-						},
-						interval: 10,
-					},
-				},
-			},
-			expectedTargets: [][]*targetgroup.Group{
-				{
-					{
-						Source:  "tp1_group1",
-						Targets: []model.LabelSet{{"__instance__": "1"}},
-					},
-					{
-						Source:  "tp1_group2",
-						Targets: []model.LabelSet{{"__instance__": "2"}},
-					},
-				},
-				{
-					{
-						Source:  "tp1_group1",
-						Targets: []model.LabelSet{{"__instance__": "3"}},
-					},
-				},
-			},
-		},
 	}
 
 	for testIndex, testCase := range testCases {

--- a/discovery/manager_test.go
+++ b/discovery/manager_test.go
@@ -649,6 +649,57 @@ func TestDiscoveryManagerSyncCalls(t *testing.T) {
 				},
 			},
 		},
+		{
+			title: "Single TP update with an empty group to check for memory leaks",
+			updates: map[string][]update{
+				"tp1": {
+					{
+						targetGroups: []targetgroup.Group{
+							{
+								Source:  "tp1_group1",
+								Targets: []model.LabelSet{{"__instance__": "1"}},
+							},
+							{
+								Source:  "tp1_group2",
+								Targets: []model.LabelSet{{"__instance__": "2"}},
+							},
+						},
+						interval: 30,
+					},
+					{
+						targetGroups: []targetgroup.Group{
+							{
+								Source:  "tp1_group1",
+								Targets: []model.LabelSet{{"__instance__": "3"}},
+							},
+							{
+								Source:  "tp1_group2",
+								Targets: nil,
+							},
+						},
+						interval: 10,
+					},
+				},
+			},
+			expectedTargets: [][]*targetgroup.Group{
+				{
+					{
+						Source:  "tp1_group1",
+						Targets: []model.LabelSet{{"__instance__": "1"}},
+					},
+					{
+						Source:  "tp1_group2",
+						Targets: []model.LabelSet{{"__instance__": "2"}},
+					},
+				},
+				{
+					{
+						Source:  "tp1_group1",
+						Targets: []model.LabelSet{{"__instance__": "3"}},
+					},
+				},
+			},
+		},
 	}
 
 	for testIndex, testCase := range testCases {

--- a/retrieval/manager.go
+++ b/retrieval/manager.go
@@ -146,20 +146,15 @@ func (m *ScrapeManager) reload(t map[string][]*targetgroup.Group) error {
 		} else {
 			existing.Sync(tgroup)
 		}
+	}
 
-		// Cleanup - check the config and cancel the scrape loops if it don't exist in the scrape config.
-		jobs := make(map[string]struct{})
-
-		for k := range m.scrapeConfigs {
-			jobs[k] = struct{}{}
-		}
-
-		for name, sp := range m.scrapePools {
-			if _, ok := jobs[name]; !ok {
-				sp.stop()
-				delete(m.scrapePools, name)
-			}
+	// Cleanup - check the config and cancel the scrape loops if it don't exist in the scrape config.
+	for name, sp := range m.scrapePools {
+		if _, ok := m.scrapeConfigs[name]; !ok {
+			sp.stop()
+			delete(m.scrapePools, name)
 		}
 	}
+
 	return nil
 }

--- a/retrieval/manager.go
+++ b/retrieval/manager.go
@@ -83,9 +83,11 @@ func (m *ScrapeManager) Stop() {
 func (m *ScrapeManager) ApplyConfig(cfg *config.Config) error {
 	done := make(chan struct{})
 	m.actionCh <- func() {
+		c := make(map[string]*config.ScrapeConfig)
 		for _, scfg := range cfg.ScrapeConfigs {
-			m.scrapeConfigs[scfg.JobName] = scfg
+			c[scfg.JobName] = scfg
 		}
+		m.scrapeConfigs = c
 		close(done)
 	}
 	<-done

